src/Control/SAMLController.php

```diff
@@ -80,194 +90,200 @@ public function acs()
+                self::class . '.ERR_SAML_ACS_FAILURE',
+                'Unfortunately we couldn\'t log you in. If this continues, please contact your I.T. department with the'
+                    . ' following reference: {ref}',
+                ['ref' => $uniqueErrorId]
+            ));
+            // We never completed to the "logged in" state, so redirect the user back to the login form to display the
+            // generic error message and reference
+            return $this->redirect(Security::login_url());
+        }
+
+        return $this->getRedirect();
+    }
+
+    private function processAuthentication(Auth $auth): void
+    {
         // Attempt to process the SAML response. If there are errors during this, log them and redirect to the generic
         // error page. Note: This does not necessarily include all SAML errors (e.g. we still need to confirm if the
         // user is authenticated after this block
+        $error = null;
+        $caughtException = null;
         try {
             $auth->processResponse();
             $error = $auth->getLastErrorReason();
         } catch (Exception $e) {
             $caughtException = $e;
+            $error = sprintf('[code: %s] %s (%s:%s)', $e->getCode(), $e->getMessage(), $e->getFile(), $e->getLine());
         }
 
         // If there was an issue with the SAML response, if it was missing or if the SAML response indicates that they
         // aren't authorised, then log the issue and provide a traceable error back to the user via the login form
-        $hasError = $caughtException || !empty($error);
-        if ($hasError || !$auth->isAuthenticated() || $this->checkForReplayAttack($auth, $uniqueErrorId)) {
-            if ($caughtException instanceof Exception) {
-                $this->getLogger()->error(sprintf(
-                    '[%s] [code: %s] %s (%s:%s)',
-                    $uniqueErrorId,
-                    $e->getCode(),
-                    $e->getMessage(),
-                    $e->getFile(),
-                    $e->getLine()
-                ));
-            }
-
-            if (!empty($error)) {
-                $this->getLogger()->error(sprintf('[%s] %s', $uniqueErrorId, $error));
-            }
+        if ($error || !$auth->isAuthenticated()) {
+            throw new AcsFailure($error ?: 'Authentication failed at IdP', 0, $caughtException);
+        }
+    }
 
-            $this->getForm()->sessionMessage(
-                _t(
-                    'SilverStripe\\SAML\\Control\\SAMLController.ERR_SAML_ACS_FAILURE',
-                    'Unfortunately we couldn\'t log you in. If this continues, please contact your I.T. department'
-                        . ' with the following reference: {ref}',
-                    ['ref' => $uniqueErrorId]
-                ),
-                ValidationResult::TYPE_ERROR
-            );
+    /**
+     * If processing reaches here, then the user is authenticated but potentially not valid. We first need to confirm
```

Comment:
This first paragraph feels like it should be a comment in the `acs()` method (e.g. so it goes alongside the other one that says "if processing reaches here the user is authenticated")

Reply:
I see now this is part of the original code - as it's my only comment I'm happy to leave this as is and merge the change :)